### PR TITLE
Fix TestPlatform.ObjectModel reference

### PR DESCRIPTION
**Pull request by mikeblakeuk**
Using net 4.7.2, dotnet test
Given we use
`AppDomain.CurrentDomain.GetAssemblies();`
The TestAdapter has a Machine TestAdapter 2.10.1 references v11 of VS Object Model. 
This throws:
```
System.IO.FileNotFoundException: [ System.IO.FileNotFoundException: Could not load file or assembly 'Microsoft.VisualStudio.TestPlatform.ObjectModel, Version=11.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified. File name: 'Microsoft.VisualStudio.TestPlatform.ObjectModel, Version=11.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' at System.Reflection.RuntimeAssembly.GetExportedTypes(RuntimeAssembly assembly, ObjectHandleOnStack retTypes) at System.Reflection.RuntimeAssembly.GetExportedTypes()
```

This ONLY happens when using dotnet / vstest console, not within the IDE

We can now remove net452

(Mono.Cecil can always be included to make the packaging logic easier)

### CODE REVIEW

**Comment by robertcoltheart** [comment]
1. We can't just remove `net452`, as that's a breaking change. Long term we want to go there, but that's a 2.0 thing.
2. I ran the scenario you described above, but I'm not seeing the same error. Can you link to a repo to help reproduce this bug, or attach a solution please?

**Comment by mikeblakeuk** [comment]
image

I don't think it's a fair test.

**Microsoft.VisualStudio.TestPlatform.ObjectModel.dll** may or may not exist.

For our test csproj, we add a reference to 
    <PackageReference Include="Machine.Specifications" Version="1.1.0-beta.3" />
    <PackageReference Include="Machine.Specifications.Runner.VisualStudio" Version="2.11.0" />
    <PackageReference Include="Machine.Specifications.Should" Version="1.0.0" />
    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.9.4" />
    
   But the Microsoft.NET.Test.Sdk doesn't copy the Microsoft.VisualStudio.TestPlatform.ObjectModel.dll in
   
   so DevOps / dotnet test fails

**Comment by robertcoltheart** [comment]
This is by design, the ObjectModel dll is part of the dotnet test harness and doesn't need to be present for the runner to function. This sounds very similar to the xunit issue over here https://github.com/xunit/xunit/issues/1852 and is a problem when a test inadvertently loads the MSpec runner dll.

**Comment by mikeblakeuk** [comment]
By net472 can use v16

**Comment by mikeblakeuk** [comment]
Just for info, this would fix the Assembly load issue.

File: src/Machine.Specifications.Runner.VisualStudio/Machine.Specifications.Runner.VisualStudio.csproj [hunk 1]
```diff
@@ -16,17 +16,17 @@
...
 
-  <PropertyGroup Condition="'$(TargetFramework)'=='netcoreapp2.1'">
+  <PropertyGroup Condition="'$(TargetFramework)' == 'netcoreapp2.1'">
     <DefineConstants>NETSTANDARD</DefineConstants>
   </PropertyGroup>
 
   <ItemGroup>
     <PackageReference Include="Mono.Cecil" Version="[0.10-*, 0.11)" />
-    <PackageReference Include="Machine.Specifications" Version="[0.11.0,2.0.0)" PrivateAssets="All" />
+    <PackageReference Include="Machine.Specifications" Version="1.*" PrivateAssets="All" />
```
Comment (robertcoltheart) [inline src/Machine.Specifications.Runner.VisualStudio/Machine.Specifications.Runner.VisualStudio.csproj:25]: Revert this please, the range above is what we want.

File: src/Machine.Specifications.Runner.VisualStudio/Machine.Specifications.Runner.VisualStudio.csproj [hunk 2]
```diff
@@ -2,7 +2,7 @@
 
   <PropertyGroup>
     <Version>0.1.0</Version>
-    <TargetFrameworks>netcoreapp2.1;net452</TargetFrameworks>
+    <TargetFrameworks>netcoreapp2.1;net452;net472</TargetFrameworks>
```
Comment (robertcoltheart) [inline src/Machine.Specifications.Runner.VisualStudio/Machine.Specifications.Runner.VisualStudio.csproj:5]: No need for net472, remove

File: src/Machine.Specifications.Runner.VisualStudio/Machine.Specifications.Runner.VisualStudio.csproj [hunk 3]
```diff
@@ -16,17 +16,17 @@
...
   <ItemGroup>
     <PackageReference Include="Mono.Cecil" Version="[0.10-*, 0.11)" />
-    <PackageReference Include="Machine.Specifications" Version="[0.11.0,2.0.0)" PrivateAssets="All" />
+    <PackageReference Include="Machine.Specifications" Version="1.*" PrivateAssets="All" />
   </ItemGroup>
 
-  <ItemGroup Condition="'$(TargetFramework)'=='netcoreapp2.1'">
-    <PackageReference Include="Microsoft.TestPlatform.ObjectModel" Version="[15.0.0,17.0.0)" PrivateAssets="All" />
+  <ItemGroup Condition="'$(TargetFramework)'!='net452'">
+    <PackageReference Include="Microsoft.TestPlatform.ObjectModel" Version="16.*" PrivateAssets="All" />
```
Comment (robertcoltheart) [inline src/Machine.Specifications.Runner.VisualStudio/Machine.Specifications.Runner.VisualStudio.csproj:29]: Let's remove the conditions and just have a single package reference for all frameworks. I was wrong earlier, we only needed to support v11 for `net35`, which technically MSpec still supports. However the min version for the VS runner is `net452` and v16 works. 

Can you also set the versions of the ObjectModel dll to be `16.6.1`? That's the version xunit is using so I feel comfortable using that version.

**Review by robertcoltheart — CHANGES_REQUESTED** [review]

File: src/Machine.Specifications.Runner.VisualStudio/Machine.Specifications.Runner.VisualStudio.csproj (hunk 3 above)
Comment (mikeblakeuk) [reply src/Machine.Specifications.Runner.VisualStudio/Machine.Specifications.Runner.VisualStudio.csproj:29]: Why not use 16.10.0? Does xunit need to be updated!

File: src/Machine.Specifications.Runner.VisualStudio/Machine.Specifications.Runner.VisualStudio.csproj (hunk 1 above)
Comment (mikeblakeuk) [reply src/Machine.Specifications.Runner.VisualStudio/Machine.Specifications.Runner.VisualStudio.csproj:25]: Shouldn't everyone be on v1 of Machine Spec?

File: src/Machine.Specifications.Runner.VisualStudio/Machine.Specifications.Runner.VisualStudio.csproj (hunk 3 above)
Comment (mikeblakeuk) [reply src/Machine.Specifications.Runner.VisualStudio/Machine.Specifications.Runner.VisualStudio.csproj]: FYI https://github.com/xunit/visualstudio.xunit/pull/273

Comment (mikeblakeuk) [reply src/Machine.Specifications.Runner.VisualStudio/Machine.Specifications.Runner.VisualStudio.csproj]: image
1. What do we build with? Product version 1.6.1 (but 16.* would be better)
2. What do we test with? Product version 1.6.10
3. What do we demand in the nuget? nothing (except Ceil)
4. What do we reference in the shipped dll? File version 15

Microsoft.VisualStudio.TestPlatform.ObjectModel, Version=15.0.0 dll, it might mean 
Microsoft.TestPlatform.ObjectModel, Version=16.x. nuget
image

Runner may or maynot include the Microsoft.VisualStudio.TestPlatform.ObjectModel.dll. If you load all assemblies, in a test, "v15" might not be found

TL;DR; we should remove v11.

Not sure we should "copy" xunit

Comment (mikeblakeuk) [reply src/Machine.Specifications.Runner.VisualStudio/Machine.Specifications.Runner.VisualStudio.csproj]: https://github.com/nunit/nunit3-vs-adapter/blob/master/src/NUnitTestAdapter/NUnit.TestAdapter.csproj#L53

File: src/Machine.Specifications.Runner.VisualStudio/Machine.Specifications.Runner.VisualStudio.csproj (hunk 1 above)
Comment (robertcoltheart) [reply src/Machine.Specifications.Runner.VisualStudio/Machine.Specifications.Runner.VisualStudio.csproj]: Not necessarily, we still maintain backwards compatibility.

File: src/Machine.Specifications.Runner.VisualStudio/Machine.Specifications.Runner.VisualStudio.csproj [hunk 4]
```diff
@@ -16,20 +16,14 @@
...
     <PackageReference Include="Mono.Cecil" Version="[0.10-*, 0.11)" />
     <PackageReference Include="Machine.Specifications" Version="[0.11.0,2.0.0)" PrivateAssets="All" />
-  </ItemGroup>
-
-  <ItemGroup Condition="'$(TargetFramework)'=='netcoreapp2.1'">
-    <PackageReference Include="Microsoft.TestPlatform.ObjectModel" Version="[15.0.0,17.0.0)" PrivateAssets="All" />
-  </ItemGroup>
-  <ItemGroup Condition="'$(TargetFramework)'=='net452'">
-    <PackageReference Include="Microsoft.TestPlatform.ObjectModel" Version="11.0.0" PrivateAssets="All" />
+    <PackageReference Include="Microsoft.TestPlatform.ObjectModel" Version="16.6.1" />
```
Comment (robertcoltheart) [inline src/Machine.Specifications.Runner.VisualStudio/Machine.Specifications.Runner.VisualStudio.csproj:26]: This is wrong, it should be private assets.

**Review by robertcoltheart — CHANGES_REQUESTED** [review]

File: src/Machine.Specifications.Runner.VisualStudio/Machine.Specifications.Runner.VisualStudio.csproj (hunk 4 above)
Comment (mikeblakeuk) [reply src/Machine.Specifications.Runner.VisualStudio/Machine.Specifications.Runner.VisualStudio.csproj]: image
how about this?

File: src/Machine.Specifications.Runner.VisualStudio/Machine.Specifications.Runner.VisualStudio.csproj [hunk 5]
```diff
@@ -40,19 +34,24 @@
...
-  <Target Name="NetFrameworkPackageItems" Condition="'$(TargetFramework)'=='net452'">
+  <Target Name="NetFrameworkPackageItems" Condition="'$(TargetFramework)' != 'netcoreapp2.1'">
     <ItemGroup>
-      <TfmSpecificPackageFile Include="*.props" PackagePath="build\$(TargetFramework)" />
-      <TfmSpecificPackageFile Include="$(OutputPath)\Machine.Specifications.dll" PackagePath="build\$(TargetFramework)\" />
-      <TfmSpecificPackageFile Include="$(OutputPath)\Machine.TestAdapter.dll" PackagePath="build\$(TargetFramework)" />
-      <TfmSpecificPackageFile Include="$(OutputPath)\Mono.*.dll" PackagePath="build\$(TargetFramework)" />
+      <TfmSpecificPackageFile Include="*.props" PackagePath="lib\$(TargetFramework)" />
+      <TfmSpecificPackageFile Include="$(OutputPath)\Machine.*.dll" PackagePath="lib\$(TargetFramework)" />
+      <TfmSpecificPackageFile Include="$(OutputPath)\Mono.*.dll" PackagePath="lib\$(TargetFramework)" />
```
Comment (robertcoltheart) [inline src/Machine.Specifications.Runner.VisualStudio/Machine.Specifications.Runner.VisualStudio.csproj:54]: Revert all this please, we want to be specific about what goes in the package. Also you've changed the path, which is the wrong path, it should be `build`.

**Review by robertcoltheart — CHANGES_REQUESTED** [review]

**Comment by mikeblakeuk** [comment]
Think this is ready now

File: src/Machine.Specifications.Runner.VisualStudio/Machine.Specifications.Runner.VisualStudio.csproj [hunk 6]
```diff
@@ -40,18 +34,23 @@
...
     </ItemGroup>
   </Target>
 
-  <Target Name="NetFrameworkPackageItems" Condition="'$(TargetFramework)'=='net452'">
+  <Target Name="NetFrameworkPackageItems" Condition="'$(TargetFramework)' != 'netcoreapp2.1'">
     <ItemGroup>
       <TfmSpecificPackageFile Include="*.props" PackagePath="build\$(TargetFramework)" />
-      <TfmSpecificPackageFile Include="$(OutputPath)\Machine.Specifications.dll" PackagePath="build\$(TargetFramework)\" />
-      <TfmSpecificPackageFile Include="$(OutputPath)\Machine.TestAdapter.dll" PackagePath="build\$(TargetFramework)" />
+      <TfmSpecificPackageFile Include="$(OutputPath)\Machine.*.dll" PackagePath="build\$(TargetFramework)" />
```
Comment (robertcoltheart) [inline src/Machine.Specifications.Runner.VisualStudio/Machine.Specifications.Runner.VisualStudio.csproj:53]: Revert pls

File: src/Machine.Specifications.Runner.VisualStudio/Machine.Specifications.Runner.VisualStudio.csproj [hunk 7]
```diff
@@ -40,18 +34,23 @@
...
+    <ItemGroup>
+      <ReferenceCopyLocalPaths Remove="@(ReferenceCopyLocalPaths)" Condition="'%(Filename)' == 'Microsoft.VisualStudio.TestPlatform.ObjectModel'" />
+    </ItemGroup>
+  </Target>
+
   <Target Name="NetCorePackageItems" Condition="'$(TargetFramework)' == 'netcoreapp2.1'">
     <ItemGroup>
       <TfmSpecificPackageFile Include="*.props" PackagePath="build\$(TargetFramework)" />
-      <TfmSpecificPackageFile Include="$(OutputPath)\*.dll" PackagePath="build\$(TargetFramework)" />
+      <TfmSpecificPackageFile Include="$(OutputPath)\Machine.*.dll" PackagePath="build\$(TargetFramework)" />
```
Comment (robertcoltheart) [inline src/Machine.Specifications.Runner.VisualStudio/Machine.Specifications.Runner.VisualStudio.csproj:46]: this too

**Review by robertcoltheart — CHANGES_REQUESTED** [review]

**Review by robertcoltheart — APPROVED** [review]

**Comment by mikeblakeuk** [comment]
Amazing, thank you
